lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,308 @@
+
+        # Creates cube list to hold lapse rate data.
+        lapse_rate_cube_list = iris.cube.CubeList([])
+
+        slices_over_realization = temperature_cube.slices_over("realization")
+        for temp_slice in slices_over_realization:
+
+            # Create slice to store lapse rate values.
+            lapse_rate_slice = temp_slice
+
+            temperature_data = temp_slice.data
+
+            # Fill sea points with NaN values. Can't use Numpy mask since not
+            # recognised by "generic_filter" function.
+            temperature_data = np.where(land_sea_mask, temperature_data,
+                                        np.nan)
+
+            # cval is value given to points outside the array.
+            fnc = fnc_class(allbuffers=all_temp_subsections)
+            generic_filter(temperature_data, fnc.filter, size=footprint_len,
+                           mode='constant', cval=np.nan)
+
+            fnc = fnc_class(allbuffers=all_orog_subsections)
+            generic_filter(orography_data, fnc.filter, size=footprint_len,
+                           mode='constant', cval=np.nan)
+
+            # height_diff_mask is True for points where the height
+            # difference between the central point and its neighbours
+            # is > max_height_diff.
+            height_diff_mask = create_height_diff_mask(
+                all_orog_subsections, ind_central_point, max_height_diff)
+
+            # Mask points with extreme height differences as NaN.
+            all_orog_subsections = np.where(height_diff_mask, np.nan,
+                                            all_orog_subsections)
+            all_temp_subsections = np.where(height_diff_mask, np.nan,
+                                            all_temp_subsections)
+
+            # Loop through both arrays and find gradient of each subsection.
+            # The gradient indicates lapse rate - save into another array.
+            lapse_rate_array = np.zeros(dataarray_size)
+
+            # TODO: This for loop is the bottleneck in the code - 1 array takes
+            # approximately 29 seconds out of a total running time of 33
+            # seconds for UKVX data. This section needs to be parallelised.
+            for i in range(dataarray_size):
+                lapse_rate_array[i] = calc_lapse_rate(
+                    all_temp_subsections[i, :], all_orog_subsections[i, :],
+                    ind_central_point)
+
+            lapse_rate_array = lapse_rate_array.reshape(dataarray_shape)
+
+            # According to the paper - any value above DALR is reset to DALR.
+            # HOWEVER the FORTRAN code also imposes a lower limit (see class
+            # docstring for code location). This will need more investigation.
+            lapse_rate_array = np.where(lapse_rate_array > DALR, DALR,
+                                        lapse_rate_array)
+            lapse_rate_array = np.where(lapse_rate_array < -3.0*DALR, DALR,
```

Comment:
We should probably add some min_lapse_rate, max_lapse_rate arguments to the class init that can be set by the user, with these values as the defaults. That way it will be easy to experiment with different limits.